File: metautils/lib/cache.c
/*
OpenIO SDS metautils
Copyright (C) 2015 OpenIO, original work as part of OpenIO Software Defined Storage

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 3.0 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library.
*/

#include "metautils.h"

/* NOOP --------------------------------------------------------------------- */

struct metautils_cache_NOOP_s
{
	const struct metautils_cache_vtable_s *vtable;
};


static void
_noop_destroy (struct metautils_cache_s *self)
{
	SLICE_FREE(struct metautils_cache_NOOP_s, self);
}

static void
_noop_put (struct metautils_cache_s *self, const char *k, const char *v)
{
	(void) self, (void) k, (void) v;
}

static void
_noop_del (struct metautils_cache_s *self, const char *k)
{
	(void) self, (void) k;
}

static gchar *
_noop_get (struct metautils_cache_s *self, const char *k)
{
	(void) self, (void) k;
	return NULL;
}

static struct metautils_cache_vtable_s vtable_NOOP =
{
	_noop_destroy, _noop_put, _noop_del, _noop_get
};

struct metautils_cache_s *
metautils_cache_make_NOOP (void)
{
	struct metautils_cache_NOOP_s *self = SLICE_NEW0 (struct metautils_cache_NOOP_s);
	self->vtable = &vtable_NOOP;
	return (struct metautils_cache_s*) self;
}

/* LRU ---------------------------------------------------------------------- */

struct metautils_cache_LRU_s;

static void _lru_destroy (struct metautils_cache_s *self);
static void _lru_put (struct metautils_cache_s *self, const char *k, const char *v);
static void _lru_del (struct metautils_cache_s *self, const char *k);
static gchar * _lru_get (struct metautils_cache_s *self, const char *k);

static struct metautils_cache_vtable_s vtable_LRU =
{
	_lru_destroy, _lru_put, _lru_del, _lru_get
};

struct metautils_cache_LRU_s
{
	const struct metautils_cache_vtable_s *vtable;
	struct lru_tree_s *lru;
};

struct metautils_cache_s *
metautils_cache_make_LRU (struct lru_tree_s *lru)
{
	EXTRA_ASSERT (lru != NULL);
	struct metautils_cache_LRU_s *self = SLICE_NEW0 (struct metautils_cache_LRU_s);
	self->vtable = &vtable_LRU;
	self->lru = lru;
	return (struct metautils_cache_s*) self;
}

static void
_lru_destroy (struct metautils_cache_s *self)
{
	struct metautils_cache_LRU_s *l = (struct metautils_cache_LRU_s*) self;
	if (!l)
		return;
	lru_tree_destroy (l->lru);
	l->lru = NULL;
	SLICE_FREE (struct metautils_cache_LRU_s, l);
}

static void
_lru_put (struct metautils_cache_s *self, const char *k, const char *v)
{
	(void) self, (void) k, (void) v;
}

static void
_lru_del (struct metautils_cache_s *self, const char *k)
{
	(void) self, (void) k;
}

static gchar *
_lru_get (struct metautils_cache_s *self, const char *k)
{
	(void) self, (void) k;
	return NULL;
}

File: oio/blob/cache.py
# Copyright (C) 2015-2017 OpenIO SAS, as part of OpenIO SDS
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 3.0 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library.


from time import time
from urlparse import urlparse

from oio.common.http_urllib3 import get_pool_manager
from oio.common.logger import get_logger
from oio.conscience.client import ConscienceClient


CACHE_TIMEOUT = 60  # in seconds


class ServiceCache(object):
    """A caching client to rawx services."""

    def __init__(self, conf, pool_manager=None):
        self._cache = dict()
        self.conf = conf
        self.pool_manager = pool_manager or get_pool_manager()
        self._client = ConscienceClient(conf=self.conf,
                                        pool_manager=self.pool_manager)
        self.logger = get_logger(conf)

    def _get_addr(self, item):
        if item in self._cache and self._cache[item]['ts'] < time():
            return self._cache[item]['addr']
        body = self._client.resolve(srv_type='rawx', service_id=item)
        self._cache[item] = {'addr': body['addr'],
                             'ts': time() + CACHE_TIMEOUT}
        return body['addr']

    def resolve(self, url):
        """
        :rtype: return resolved url of a rawx using Service-ID
        """
        # FIXME(mb) some tests doesn't put scheme, should fix tests
        if not url.startswith('http://'):
            url = "http://" + url
        res = urlparse(url)
        if res.port == 80 or res.port is None:
            service_id_or_host = urlparse(url).hostname
            host = self._get_addr(service_id_or_host)
            if host != service_id_or_host:
                url = res.scheme + "://" + host + res.path
                if res.query:
                    url += '?' + res.query
        return url
